python/ml4ir/data/tfrecord_reader.py

```diff
@@ -79,72 +88,102 @@ def _parse_sequence_example_fn(sequence_example_proto):
 
             features_dict[feature_node_name] = feature_tensor
 
+        # Define mask to identify padded records
+        if required_only and not feature_config.get_rank("serving_info")["required"]:
+            """
+            Define dummy mask if the rank field is not a required field for serving
+
+            NOTE:
+            This masks all max_num_records as 1 as there is no real way to know
+            the number of records in the query. There is no predefined required field,
+            and hence we would need to do a full pass of all features to find the record shape.
+            This approach might be unstable if different features have different shapes.
+
+            Hence we just mask all records
+            """
+            features_dict["mask"] = tf.constant(
+                value=1, shape=[max_num_records], dtype=feature_config.get_rank("dtype")
+            )
+            num_records = tf.constant(max_num_records, dtype=tf.int64)
+        else:
+            # Typically used at training time, to pad/clip to a fixed number of records per query
+
+            # Use rank as a reference tensor to infer shape/num_records in query
```

Comment:
@jakemannix I use `rank` as a reference tensor to infer the number of records in the query and mask it.